lib/host/wasi_nn/wasinnfunc.cpp

```diff
@@ -3,49 +3,482 @@
+    spdlog::error("[WASI-NN] init_execution_context: Graph Id does not exist");
+    return WASINN::InvalidArgument;
+  }
+
+  if (Ctx.GraphBackends[GraphId] == Ctx.BackendsMapping.at("OpenVINO")) {
+#ifdef WASMEDGE_WASINN_BUILD_OPENVINO
+    IEStatusCode Status;
+    WASINN::OpenVINOSession *Session = new WASINN::OpenVINOSession();
+    Session->ExeNetwork = Ctx.OpenVINOExecutions[GraphId];
+    Session->Network = Ctx.OpenVINONetworks[GraphId];
+    if (Session->ExeNetwork == nullptr || Session->Network == nullptr) {
+      spdlog::error("[WASI-NN] Model for Graph:{} is empty!", GraphId);
+      delete Session;
+      return WASINN::MissingMemory;
+    }
+    Status = ie_exec_network_create_infer_request(Session->ExeNetwork,
+                                                  &(Session->InferRequest));
+
+    if (Status != IEStatusCode::OK || Session->InferRequest == nullptr) {
+      spdlog::error("[WASI-NN] Unable to create openvino session");
+      return WASINN::Busy;
+    }
+
+    Ctx.ExecutionsNum = Ctx.OpenVINOInfers.size();
+    Ctx.OpenVINOInfers.push_back(Session);
+    Ctx.GraphContextBackends.push_back(GraphId);
+    *Context = Ctx.ExecutionsNum;
+
+    return WASINN::Success;
+#else
+    spdlog::error("[WASI-NN] OpenVINO backend is not built. define "
+                  "-DWASMEDGE_WASINN_BUILD_OPENVINO "
+                  "to build it.");
+#endif
+  } else {
+    spdlog::error("[WASI-NN] Current backend is not supported.");
+  }
+  return WASINN::InvalidArgument;
 }
 
-Expect<uint32_t> WasiNNSetInput::body(Runtime::Instance::MemoryInstance *MemInst
-                                      [[maybe_unused]],
-                                      uint32_t Context [[maybe_unused]],
-                                      uint32_t Index [[maybe_unused]],
-                                      uint32_t TensorPtr [[maybe_unused]]) {
-  return 0;
+Expect<WASINN::NNErrNo>
+WasiNNSetInput::body(Runtime::Instance::MemoryInstance *MemInst,
+                     uint32_t Context, uint32_t Index [[maybe_unused]],
+                     uint32_t TensorPtr [[maybe_unused]]) {
+  if (MemInst == nullptr) {
+    return Unexpect(ErrCode::ExecutionFailed);
+  }
+
+  if (Ctx.GraphContextBackends.size() <= Context) {
+    spdlog::error("[WASI-NN] set_input: Execution Context does not exist");
+    return WASINN::InvalidArgument;
+  }
+  if (Ctx.GraphContextBackends[Context] == Ctx.BackendsMapping.at("OpenVINO")) {
+#ifdef WASMEDGE_WASINN_BUILD_OPENVINO
+    IEStatusCode Status;
+    WASINN::OpenVINOSession *Session = Ctx.OpenVINOInfers[Context];
+    ie_network_t *Network = Session->Network;
+    ie_infer_request_t *InferRequest = Session->InferRequest;
+    ie_blob_t *InputBlob = nullptr;
+    ie_blob_buffer_t BlobCBuffer;
+    int BlobSize;
+    char *InputName = nullptr;
+
+    if (Network == nullptr || InferRequest == nullptr) {
+      spdlog::error("[WASI-NN] The founded openvino session is empty");
+      delete Session;
```

Comment:
What will the `Ctx.OpenVINOInfers[Context]` be if you delete the `Session` here?

Reply:
Thanks a lot. That might be a bug. Once the `Session` is deleted, the `Ctx.OpenVINOInfers[Context]` should be set to `nullptr`.